src/serde/ser.rs

```diff
@@ -0,0 +1,328 @@
+    }
+
+    fn serialize_tuple_struct(
+        self,
+        _name: &'static str,
+        _len: usize,
+    ) -> Result<Self::SerializeTupleStruct, Self::Error> {
+        Ok(TupleStructSerializer { fields: vec![] })
+    }
+
+    fn serialize_tuple_variant(
+        self,
+        _name: &'static str,
+        _variant_index: u32,
+        variant: &'static str,
+        _len: usize,
+    ) -> Result<Self::SerializeTupleVariant, Self::Error> {
+        Ok(TupleVariantSerializer {
+            variant,
+            fields: vec![],
+        })
+    }
+
+    fn serialize_map(self, _len: Option<usize>) -> Result<Self::SerializeMap, Self::Error> {
+        Ok(MapSerializer {
+            key: None,
+            items: vec![],
+        })
+    }
+
+    fn serialize_struct(
+        self,
+        _name: &'static str,
+        _len: usize,
+    ) -> Result<Self::SerializeStruct, Self::Error> {
+        Ok(StructSerializer { fields: vec![] })
+    }
+
+    fn serialize_struct_variant(
+        self,
+        _name: &'static str,
+        _variant_index: u32,
+        variant: &'static str,
+        _len: usize,
+    ) -> Result<Self::SerializeStructVariant, Self::Error> {
+        Ok(StructVariantSerializer {
+            variant,
+            fields: vec![],
+        })
+    }
+}
+
+pub(crate) struct SeqSerializer {
+    items: Vec<Value>,
+}
+
+impl<'a> ser::SerializeTuple for SeqSerializer {
+    type Ok = Value;
+    type Error = Error;
+
+    fn serialize_element<T: ?Sized + Serialize>(&mut self, value: &T) -> Result<(), Self::Error> {
+        self.items.push(value.serialize(&mut Serializer {})?);
+        Ok(())
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        Ok(Value::Array(self.items))
+    }
+}
+
+/// The implementation of `SerializeSeq` just forwards to `SerializeTuple`.
+impl<'a> ser::SerializeSeq for SeqSerializer {
+    type Ok = Value;
+    type Error = Error;
+
+    fn serialize_element<T: ?Sized + Serialize>(&mut self, value: &T) -> Result<(), Self::Error> {
+        <Self as ser::SerializeTuple>::serialize_element(self, value)
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        <Self as ser::SerializeTuple>::end(self)
+    }
+}
+
+pub(crate) struct StructVariantSerializer {
+    variant: &'static str,
+    fields: Vec<(Value, Value)>,
+}
+impl<'a> ser::SerializeStructVariant for StructVariantSerializer {
+    type Ok = Value;
+    type Error = Error;
+
+    fn serialize_field<T: ?Sized + Serialize>(
+        &mut self,
+        key: &'static str,
+        value: &T,
+    ) -> Result<(), Self::Error> {
+        let mut s = Serializer {};
+        self.fields
+            .push((key.serialize(&mut s)?, value.serialize(&mut s)?));
+        Ok(())
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        // Normalization (sorting) also happens inside sk_cbor, so this is
+        // technically redundant, but oasis_cbor (non-serde) does the same,
```

Comment:
The reason for this additional sorting is to be able to decode from intermediate `Value` types as otherwise `cbor::from_value` may not work (since sorting is mandatory in canonical CBOR) on values produced by `cbor::to_value`.

Reply:
I agree with your approach, let's fix decode and get rid of the value sorts in encoders.